pd/src/app.rs

```diff
@@ -443,24 +452,63 @@ impl App {
 
                 let mut next_rates = Vec::new();
                 let mut next_validator_statuses = Vec::new();
+
+                // this is a bit complicated: because we're in the EndBlock phase, and the
+                // delegations in this block have not yet been committed, we have to combine
+                // the delegations in pending_block with the ones already committed to the
+                // state. otherwise the delegations committed in the epoch threshold block
+                // would be lost.
+                let mut committed_delegation_changes =
+                    state.delegation_changes(prev_epoch.index).await?;
+                for (id_key, delta) in &pending_block.lock().unwrap().delegation_changes {
+                    let _ = *committed_delegation_changes
+                        .entry(id_key.clone())
+                        .and_modify(|change| *change += delta)
+                        .or_insert(*delta);
+                }
+
                 for current_rate in &current_rates {
                     let identity_key = current_rate.identity_key.clone();
 
                     let funding_streams = state.funding_streams(identity_key.clone()).await?;
 
                     let next_rate = current_rate.next(&next_base_rate, funding_streams);
 
-                    let delegation_token_supply = state
+                    // TODO: if a validator isn't part of the consensus set, should we ignore them
+                    // and not update their rates?
```

Comment:
Yeah, currently we don't have a notion of validator state, but when we do, we want to split into two paths:

- for validators in the consensus set, use the base rate data and funding streams to compute rate updates (the current code);
- for validators not in the consensus set, roll forward their existing rates, changing only the epoch index.

Maybe `RateData::next` should be split into `RateData::next_active` and `RateData::next_inactive`?